### PR TITLE
Feature/update user

### PR DESCRIPTION
**Pull request by mwnicks**
### What

Added the update user endpoint

### How to review

Ensure all unit and component tests are passing.
Send a request to PUT '/v1/users/{user_id}' with user_id = ae54bd71-6099-4364-acd3-519a64202679 and request body:
- forename = a new forename for the user
- lastname = a new surname for the user
- status = CONFIRMED

You should get a 200 response with the response body like:
{
       "id": "ae54bd71-6099-4364-acd3-519a64202679",
       "forename": <new forename>,
       "lastname": <new lastname>,
       "email": "matt.nicks@ons.gov.uk",
       "groups": [],
       "status": "CONFIRMED"
}

### Who can review

Any one but me

### CODE REVIEW

File: api/users.go [hunk 1]
```diff
@@ -112,6 +112,60 @@ func (api *API) GetUserHandler(ctx context.Context, w http.ResponseWriter, req *
...
+		return nil, models.NewErrorResponse(validationErrs, http.StatusBadRequest, nil)
+	}
+
+	userRequest := user.BuildUpdateUserRequest(api.UserPoolId)
+
+	_, err = api.CognitoClient.AdminUpdateUserAttributes(userRequest)
+	if err != nil {
+		responseErr := models.NewCognitoError(ctx, err, "AdminUpdateUserAttributes request from update user endpoint")
+		if responseErr.Code == models.UserNotFoundError {
+			return nil, models.NewErrorResponse([]error{responseErr}, http.StatusNotFound, nil)
```
Comment (daiLlew) [inline api/users.go]: May as well create a var for this: `[]error{responseErr}` seeing as you always need it in that format just make its easier to read.

File: api/users.go [hunk 2]
```diff
@@ -112,6 +112,60 @@ func (api *API) GetUserHandler(ctx context.Context, w http.ResponseWriter, req *
...
+		} else if responseErr.Code == models.InvalidFieldError {
+			return nil, models.NewErrorResponse([]error{responseErr}, http.StatusBadRequest, nil)
+		} else {
+			return nil, models.NewErrorResponse([]error{responseErr}, http.StatusInternalServerError, nil)
+		}
```
Comment (daiLlew) [inline api/users.go]: I don't think the `else` is required here, same is achieved if you just fall through to this return statement.

File: api/users.go [hunk 3]
```diff
@@ -112,6 +112,60 @@ func (api *API) GetUserHandler(ctx context.Context, w http.ResponseWriter, req *
...
+	}
+
+	userDetailsRequest := user.BuildAdminGetUserRequest(api.UserPoolId)
+	userDetailsResponse, err := api.CognitoClient.AdminGetUser(userDetailsRequest)
+	if err != nil {
+		responseErr := models.NewCognitoError(ctx, err, "AdminGetUser request from update user endpoint")
+		return nil, models.NewErrorResponse([]error{responseErr}, http.StatusInternalServerError, nil)
```
Comment (daiLlew) [inline api/users.go:156]: This is probably beyond the scope of this PR but I am thinking adding a new func to make this syntax a bit nice might be a good idea:
Atm we have:
```
func NewErrorResponse(errors []error, statusCode int, headers map[string]string) *ErrorResponse {}
```
But that requires you to do the following all the time when you only have 1 error.
```
[]error{responseErr}
```
Its driving my OCD wild 😂  We could add another "overloaded func":
```
func NewErrorResponse(statusCode int, headers map[string]string, errors... error) *ErrorResponse {}
```
Which would be a bit nicer to use. Again totally beyond the scope of this PR.

File: api/users_test.go [hunk 4]
```diff
@@ -380,6 +380,156 @@ func TestGetUserHandler(t *testing.T) {
...
+				http.StatusInternalServerError,
+			},
+		}
+
+		for _, tt := range adminCreateUsersTests {
+			m.AdminUpdateUserAttributesFunc = tt.updateUserFunction
+			m.AdminGetUserFunc = tt.getUserFunction
+
+			postBody := map[string]interface{}{"forename": tt.userForename, "lastname": lastname, "status": status, "role_type": roleType}
+			body, _ := json.Marshal(postBody)
```
Comment (daiLlew) [inline api/users_test.go]: Should check this error, can do a `So(err, ShouldBeNil)` which will fail the test.

File: api/users_test.go [hunk 5]
```diff
@@ -380,6 +380,156 @@ func TestGetUserHandler(t *testing.T) {
...
+			r := httptest.NewRequest(http.MethodGet, userEndPoint, bytes.NewReader(body))
+
+			successResponse, errorResponse := api.UpdateUserHandler(ctx, w, r)
+
+			// Check whether testing a success or error case
+			if tt.httpResponse > 399 {
```
Comment (daiLlew) [inline api/users_test.go]: this is relying on the tests returning the right responses for each test case. Is there a way we can add the "expectation" to the test case struct? That way if the test returns an incorrect response the expectation defined in the test case will ensure the test fails appropriately.

Comment (mwnicks) [reply api/users_test.go]: This was actually relying on response code a set in the test case struct to determine which set of assertions to user for the test case.
But I agree having the assertions as part of the test case struct makes it cleaner and easier to follow.

File: api/users.go (hunk 3 above)
Comment (mwnicks) [reply api/users.go:156]: This has actually been bugging me too.
The suggested approach is definitely cleaner.
I'll open a separate PR to refactor this across the service

File: api/api.go [hunk 6]
```diff
@@ -69,6 +69,7 @@ func Setup(ctx context.Context, r *mux.Router, cognitoClient cognito.Client, use
 	r.HandleFunc("/v1/users", contextAndErrors(api.CreateUserHandler)).Methods("POST")
 	r.HandleFunc("/v1/users", contextAndErrors(api.ListUsersHandler)).Methods("GET")
 	r.HandleFunc("/v1/users/{id}", contextAndErrors(api.GetUserHandler)).Methods("GET")
+	r.HandleFunc("/v1/users/{id}", contextAndErrors(api.UpdateUserHandler)).Methods("PUT")
```
Comment (mike-kelly-methods) [inline api/api.go]: I'm thinking Matt, it'd probably be good to get some consistency by replacing the inlining of the http verb with `net/http` constants as is the case in other areas of the codebase:

`import net/http`
...
...
...
`r.HandleFunc("/v1/users/{id}", contextAndErrors(api.UpdateUserHandler)).Methods(http.MethodPut)`

**Review by daiLlew — APPROVED** [review]

File: api/api_test.go [hunk 7]
```diff
@@ -39,6 +39,7 @@ func TestSetup(t *testing.T) {
 			So(hasRoute(api.Router, "/v1/users", "POST"), ShouldBeTrue)
 			So(hasRoute(api.Router, "/v1/users", "GET"), ShouldBeTrue)
 			So(hasRoute(api.Router, "/v1/users/{id}", "GET"), ShouldBeTrue)
+			So(hasRoute(api.Router, "/v1/users/{id}", "PUT"), ShouldBeTrue)
```
Comment (mike-kelly-methods) [inline api/api_test.go:42]: Not a biggy - but we should use the `net/http` constants here too.

**Review by mike-kelly-methods — APPROVED** [review]
👍

**Pull request merged**